File: Python/trapping_rainwater.py
class Solution:
    def trap(self, height: List[int]) -> int:
        total_vol, l, r, lm, rm = 0, 0, len(height) -1, height[0], height[-1]
        while l< r:
            if lm < rm:
                l += 1
                lm = max(lm, height[l])
                total_vol += lm - height[l]

            else:
                r -= 1
                rm = max(rm, height[r])
                total_vol += rm - height[r]
        
        return total_vol

        

File: Cpp/trapping_rainwater.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        if (height.size() == 0) return 0;
        int l,r, max_left, max_right, total_vol;
        l = total_vol = 0;
        r = height.size()-1;
        max_left = height[l]; 
        max_right = height[r];

        while(l < r)
        {
            if (max_left < max_right)
            {
                l++;
                max_left = max(max_left, height[l]);
                total_vol += max_left - height[l];
            }

            else
            {
                r--;
                max_right = max(max_right, height[r]);
                total_vol += max_right - height[r];

            }
        }
        return total_vol;
    }
};